File: back/runtimes/ai/rt-agents/src/agent/providers/claude.ts
import Anthropic from "@anthropic-ai/sdk";
import type { LLMProvider, LLMProviderChatParams, ProviderStreamEvent } from "./base";
import type { AgentMessage } from "../core/types";

let client: Anthropic | null = null;

function getClient(apiKey: string): Anthropic {
  if (!client) {
    client = new Anthropic({ apiKey });
  }
  return client;
}

export class ClaudeProvider implements LLMProvider {
  name = "anthropic";
  private client: Anthropic;

  constructor(apiKey: string) {
    this.client = getClient(apiKey);
  }

  async *chat(params: LLMProviderChatParams): AsyncGenerator<ProviderStreamEvent> {
    const messages = this.convertMessages(params.messages);

    const tools = params.tools?.map((t) => ({
      name: t.name,
      description: t.description,
      input_schema: t.parameters as Anthropic.Tool.InputSchema,
    }));

    const requestParams: any = {
      model: params.model,
      max_tokens: params.maxTokens,
      temperature: params.temperature,
      messages,
      stream: true,
    };
    if (params.systemPrompt) requestParams.system = params.systemPrompt;
    if (tools && tools.length > 0) requestParams.tools = tools;

    const stream = await this.client.messages.create(requestParams);

    // Track tool calls by content block index
    const blockToolIds: Map<number, string> = new Map();
    const blockToolNames: Map<number, string> = new Map();
    const blockToolArgs: Map<number, string> = new Map();

    for await (const event of stream as any) {
      if (event.type === "content_block_start") {
        const block = event.content_block;
        if (block?.type === "text" && block.text) {
          yield { type: "text_delta", content: block.text };
        }
        if (block?.type === "tool_use") {
          blockToolIds.set(event.index, block.id);
          blockToolNames.set(event.index, block.name);
          blockToolArgs.set(event.index, "");
          yield { type: "tool_call_start", id: block.id, name: block.name };
        }
      }

      if (event.type === "content_block_delta") {
        if (event.delta?.type === "text_delta") {
          yield { type: "text_delta", content: event.delta.text || "" };
        }
        if (event.delta?.type === "input_json_delta") {
          const idx = event.index;
          const id = blockToolIds.get(idx);
          if (id !== undefined) {
            const current = blockToolArgs.get(idx) || "";
            blockToolArgs.set(idx, current + (event.delta.partial_json || ""));
            yield { type: "tool_call_delta", id, argsJson: event.delta.partial_json || "" };
          }
        }
      }

      if (event.type === "content_block_stop") {
        const idx = event.index;
        const id = blockToolIds.get(idx);
        if (id !== undefined) {
          const argsJson = blockToolArgs.get(idx) || "{}";
          const name = blockToolNames.get(idx) || "";
          let args: Record<string, unknown> = {};
          try {
            args = JSON.parse(argsJson);
          } catch {}
          yield { type: "tool_call_end", id, name, args };
          blockToolIds.delete(idx);
          blockToolNames.delete(idx);
          blockToolArgs.delete(idx);
        }
      }

      if (event.type === "message_delta") {
        yield {
          type: "message_complete",
          finishReason: event.delta?.stop_reason || "end_turn",
          usage: {
            input: event.usage?.input_tokens || 0,
            output: event.usage?.output_tokens || 0,
          },
        };
      }
    }
  }

  private convertMessages(messages: AgentMessage[]): Anthropic.MessageParam[] {
    return messages
      .filter((m) => m.role !== "system")
      .map((msg): Anthropic.MessageParam => {
        if (msg.role === "assistant" && msg.toolCalls?.length) {
          const content: Anthropic.ContentBlock[] = [];
          if (msg.content) {
            content.push({ type: "text", text: msg.content } as any);
          }
          for (const tc of msg.toolCalls) {
            content.push({
              type: "tool_use",
              id: tc.id,
              name: tc.name,
              input: tc.args,
            } as any);
          }
          return { role: "assistant", content };
        }

        if (msg.role === "tool") {
          return {
            role: "user",
            content: [
              {
                type: "tool_result",
                tool_use_id: msg.toolCallId,
                content: msg.content,
              } as any,
            ],
          };
        }

        return {
          role: msg.role === "user" ? "user" : "assistant",
          content: msg.content,
        };
      });
  }
}

File: native/runtime/src/llm/claude.zig
//! Anthropic messages provider (non-streaming). Wire dialect mirrors the TS
//! ClaudeProvider (rt-assistant/impls/providers/claude.ts): system prompt is a
//! top-level field, tool calls/results travel as content blocks.

const std = @import("std");
const provider = @import("provider.zig");
const http = @import("http.zig");

pub const default_base_url = "https://api.anthropic.com/v1";
const anthropic_version = "2023-06-01";

pub const Config = struct {
    api_key: []const u8,
    base_url: []const u8,
};

pub fn make(cfg: *const Config) provider.Provider {
    return .{ .name = "claude", .ctx = @constCast(cfg), .complete = complete };
}

fn complete(ctx: *anyopaque, env: provider.CallEnv, req: provider.ChatRequest) anyerror!provider.Reply {
    const cfg: *const Config = @ptrCast(@alignCast(ctx));
    const a = env.alloc;

    var body = try std.ArrayList(u8).initCapacity(a, 1024);
    defer body.deinit(a);

    try body.appendSlice(a, "{\"model\":");
    try provider.appendJsonStr(&body, a, req.model);
    const budget = try std.fmt.allocPrint(a, ",\"max_tokens\":{d}", .{req.max_tokens});
    try body.appendSlice(a, budget);
    if (req.temperature) |t| {
        const temp = try std.fmt.allocPrint(a, ",\"temperature\":{d}", .{t});
        try body.appendSlice(a, temp);
    }
    if (try provider.joinSystem(a, req.messages)) |system| {
        try body.appendSlice(a, ",\"system\":");
        try provider.appendJsonStr(&body, a, system);
    }

    try body.appendSlice(a, ",\"messages\":[");
    var emitted: usize = 0;
    for (req.messages, 0..) |m, idx| {
        const role = provider.strField(m, "role") orelse
            return provider.errReply(a, "claude: message {d}: missing role", .{idx});
        if (std.mem.eql(u8, role, "system")) continue;
        if (emitted > 0) try body.appendSlice(a, ",");
        emitted += 1;
        try appendMessage(&body, a, m, role);
    }
    try body.appendSlice(a, "]");

    if (req.tools.len > 0) {
        try body.appendSlice(a, ",\"tools\":[");
        for (req.tools, 0..) |t, i| {
            if (i > 0) try body.appendSlice(a, ",");
            const name = provider.strField(t, "name") orelse "";
            const description = provider.strField(t, "description") orelse "";
            try body.appendSlice(a, "{\"name\":");
            try provider.appendJsonStr(&body, a, name);
            try body.appendSlice(a, ",\"description\":");
            try provider.appendJsonStr(&body, a, description);
            try body.appendSlice(a, ",\"input_schema\":");
            if (provider.field(t, "parameters")) |params| {
                try provider.appendValue(&body, a, params);
            } else {
                try body.appendSlice(a, "{\"type\":\"object\",\"properties\":{}}");
            }
            try body.appendSlice(a, "}");
        }
        try body.appendSlice(a, "]");
    }
    try body.appendSlice(a, "}");

    const url = try std.fmt.allocPrint(a, "{s}/messages", .{cfg.base_url});
    const headers = [_]std.http.Header{
        .{ .name = "content-type", .value = "application/json" },
        .{ .name = "x-api-key", .value = cfg.api_key },
        .{ .name = "anthropic-version", .value = anthropic_version },
    };

    const res = try http.postJson(env.client, a, url, &headers, body.items);
    if (res.status < 200 or res.status >= 300)
        return provider.errReply(a, "claude HTTP {d}: {s}", .{ res.status, res.body });

    return parseResponse(a, req.model, res.body);
}

fn appendMessage(body: *std.ArrayList(u8), a: std.mem.Allocator, m: std.json.Value, role: []const u8) !void {
    const content = provider.strField(m, "content") orelse "";

    // Tool result -> a user turn holding one tool_result block.
    if (std.mem.eql(u8, role, "tool")) {
        const call_id = provider.strField(m, "toolCallId") orelse "";
        try body.appendSlice(a, "{\"role\":\"user\",\"content\":[{\"type\":\"tool_result\",\"tool_use_id\":");
        try provider.appendJsonStr(body, a, call_id);
        try body.appendSlice(a, ",\"content\":");
        try provider.appendJsonStr(body, a, content);
        try body.appendSlice(a, "}]}");
        return;
    }

    // Assistant turn with tool calls -> text + tool_use blocks.
    const calls = if (std.mem.eql(u8, role, "assistant")) provider.arrField(m, "toolCalls") else null;
    if (calls != null and calls.?.len > 0) {
        try body.appendSlice(a, "{\"role\":\"assistant\",\"content\":[");
        var first = true;
        if (content.len > 0) {
            try body.appendSlice(a, "{\"type\":\"text\",\"text\":");
            try provider.appendJsonStr(body, a, content);
            try body.appendSlice(a, "}");
            first = false;
        }
        for (calls.?) |tc| {
            if (!first) try body.appendSlice(a, ",");
            first = false;
            try body.appendSlice(a, "{\"type\":\"tool_use\",\"id\":");
            try provider.appendJsonStr(body, a, provider.strField(tc, "id") orelse "");
            try body.appendSlice(a, ",\"name\":");
            try provider.appendJsonStr(body, a, provider.strField(tc, "name") orelse "");
            try body.appendSlice(a, ",\"input\":");
            if (provider.field(tc, "args")) |args| {
                try provider.appendValue(body, a, args);
            } else {
                try body.appendSlice(a, "{}");
            }
            try body.appendSlice(a, "}");
        }
        try body.appendSlice(a, "]}");
        return;
    }

    const wire_role = if (std.mem.eql(u8, role, "assistant")) "assistant" else "user";
    try body.appendSlice(a, "{\"role\":\"");
    try body.appendSlice(a, wire_role);
    try body.appendSlice(a, "\",\"content\":");
    try provider.appendJsonStr(body, a, content);
    try body.appendSlice(a, "}");
}

fn parseResponse(a: std.mem.Allocator, model: []const u8, raw: []const u8) !provider.Reply {
    const root = std.json.parseFromSliceLeaky(std.json.Value, a, raw, .{}) catch
        return provider.errReply(a, "claude: malformed response: {s}", .{raw});

    const blocks = provider.arrField(root, "content") orelse
        return provider.errReply(a, "claude: response without content: {s}", .{raw});

    var text = try std.ArrayList(u8).initCapacity(a, 0);
    defer text.deinit(a);
    var tool_calls = try std.ArrayList(provider.ToolCall).initCapacity(a, 0);
    defer tool_calls.deinit(a);

    for (blocks) |block| {
        const kind = provider.strField(block, "type") orelse continue;
        if (std.mem.eql(u8, kind, "text")) {
            try text.appendSlice(a, provider.strField(block, "text") orelse "");
        } else if (std.mem.eql(u8, kind, "tool_use")) {
            const args_json = if (provider.field(block, "input")) |input|
                try std.json.Stringify.valueAlloc(a, input, .{})
            else
                "{}";
            try tool_calls.append(a, .{
                .id = provider.strField(block, "id") orelse "",
                .name = provider.strField(block, "name") orelse "",
                .args_json = args_json,
            });
        }
    }

    const usage = provider.field(root, "usage");
    const body = try provider.encodeCompletion(a, "claude", model, .{
        .text = text.items,
        .tool_calls = tool_calls.items,
        .finish_reason = provider.strField(root, "stop_reason") orelse "end_turn",
        .usage_input = if (usage) |u| provider.intField(u, "input_tokens") orelse 0 else 0,
        .usage_output = if (usage) |u| provider.intField(u, "output_tokens") orelse 0 else 0,
    });
    return .{ .ok = true, .body = body };
}
